File: backend/routes/export.js
const express = require("express");
const router = express.Router();
const { isAuthenticated } = require("./auth");
const archiver = require("archiver");
const path = require("path");
const fs = require("fs");
const { exec } = require("child_process");
const { promisify } = require("util");

const execPromise = promisify(exec);

// Patterns to exclude from export
const EXCLUDE_PATTERNS = [
    'node_modules',
    '.git',
    '.env',
    'dist',
    'build',
    '.next',
    '__pycache__',
    '*.log',
    '.cache',
    'coverage',
    '.nyc_output',
    'vendor',  // PHP composer
    'venv',    // Python virtual env
    '.venv'
];

// Get PM2 process working directory
async function getPm2ProcessPath(processName) {
    try {
        const { stdout } = await execPromise(`pm2 jlist`);
        const processes = JSON.parse(stdout);
        const proc = processes.find(p => p.name === processName);

        if (proc && proc.pm2_env && proc.pm2_env.pm_cwd) {
            return proc.pm2_env.pm_cwd;
        }
        return null;
    } catch (e) {
        console.error("Error getting PM2 path:", e);
        return null;
    }
}

// Get Docker container working directory (mount paths)
async function getDockerContainerMounts(containerId) {
    try {
        const { stdout } = await execPromise(`docker inspect ${containerId} --format "{{json .Mounts}}"`);
        const mounts = JSON.parse(stdout);

        // Get source paths from bind mounts
        const paths = mounts
            .filter(m => m.Type === 'bind')
            .map(m => m.Source);

        return paths.length > 0 ? paths[0] : null;
    } catch (e) {
        console.error("Error getting Docker mounts:", e);
        return null;
    }
}

// Check if path should be excluded
function shouldExclude(filePath) {
    const basename = path.basename(filePath);
    return EXCLUDE_PATTERNS.some(pattern => {
        if (pattern.includes('*')) {
            // Simple wildcard matching
            const regex = new RegExp('^' + pattern.replace('*', '.*') + '$');
            return regex.test(basename);
        }
        return basename === pattern;
    });
}

// Export PM2 project
router.get("/pm2/:name", isAuthenticated, async (req, res) => {
    const processName = req.params.name;

    try {
        const projectPath = await getPm2ProcessPath(processName);

        if (!projectPath || !fs.existsSync(projectPath)) {
            return res.status(404).json({
                success: false,
                error: "Project path not found. Make sure the PM2 process is running."
            });
        }

        const zipName = `${processName}-${Date.now()}.zip`;

        res.setHeader('Content-Type', 'application/zip');
        res.setHeader('Content-Disposition', `attachment; filename="${zipName}"`);

        const archive = archiver('zip', { zlib: { level: 5 } });

        archive.on('error', (err) => {
            console.error("Archive error:", err);
            if (!res.headersSent) {
                res.status(500).json({ success: false, error: err.message });
            }
        });

        archive.pipe(res);

        // Add files recursively, excluding patterns
        archive.glob('**/*', {
            cwd: projectPath,
            ignore: EXCLUDE_PATTERNS,
            dot: true
        });

        await archive.finalize();

    } catch (error) {
        console.error("Export error:", error);
        if (!res.headersSent) {
            res.status(500).json({ success: false, error: error.message });
        }
    }
});

// Export Docker project (from mount path)
router.get("/docker/:id", isAuthenticated, async (req, res) => {
    const containerId = req.params.id;

    try {
        const projectPath = await getDockerContainerMounts(containerId);

        if (!projectPath || !fs.existsSync(projectPath)) {
            return res.status(404).json({
                success: false,
                error: "No bind mount found or path doesn't exist. Container might be using volumes instead of bind mounts."
            });
        }

        const containerName = containerId.substring(0, 12);
        const zipName = `docker-${containerName}-${Date.now()}.zip`;

        res.setHeader('Content-Type', 'application/zip');
        res.setHeader('Content-Disposition', `attachment; filename="${zipName}"`);

        const archive = archiver('zip', { zlib: { level: 5 } });

        archive.on('error', (err) => {
            console.error("Archive error:", err);
            if (!res.headersSent) {
                res.status(500).json({ success: false, error: err.message });
            }
        });

        archive.pipe(res);

        // Add files recursively, excluding patterns
        archive.glob('**/*', {
            cwd: projectPath,
            ignore: EXCLUDE_PATTERNS,
            dot: true
        });

        await archive.finalize();

    } catch (error) {
        console.error("Export error:", error);
        if (!res.headersSent) {
            res.status(500).json({ success: false, error: error.message });
        }
    }
});

module.exports = router;

File: be/internal/handlers/export.go
package handlers

import (
	"archive/zip"
	"context"
	"encoding/json"
	"io"
	"io/fs"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/kaysa/home-panel/internal/httpx"
)

// Export ports backend/routes/export.js: stream a project directory as a zip,
// excluding heavy/secret paths. Source path comes from PM2 cwd or a Docker bind mount.
type Export struct{}

// excludedDirs / excludedFiles mirror EXCLUDE_PATTERNS in export.js.
var excludedDirs = map[string]bool{
	"node_modules": true, ".git": true, "dist": true, "build": true,
	".next": true, "__pycache__": true, ".cache": true, "coverage": true,
	".nyc_output": true, "vendor": true, "venv": true, ".venv": true,
	".idea": true, ".vscode": true,
}

func isExcluded(rel string) bool {
	for _, seg := range strings.Split(filepath.ToSlash(rel), "/") {
		if excludedDirs[seg] {
			return true
		}
	}
	base := filepath.Base(rel)
	return base == ".env" || strings.HasSuffix(base, ".log")
}

func (Export) pm2Path(ctx context.Context, name string) string {
	out, err := exec.CommandContext(ctx, "pm2", "jlist").Output()
	if err != nil {
		return ""
	}
	var procs []struct {
		Name   string `json:"name"`
		PM2Env struct {
			PmCwd string `json:"pm_cwd"`
		} `json:"pm2_env"`
	}
	if json.Unmarshal(out, &procs) != nil {
		return ""
	}
	for _, p := range procs {
		if p.Name == name {
			return p.PM2Env.PmCwd
		}
	}
	return ""
}

func (Export) dockerMount(ctx context.Context, id string) string {
	out, err := exec.CommandContext(ctx, "docker", "inspect", id, "--format", "{{json .Mounts}}").Output()
	if err != nil {
		return ""
	}
	var mounts []struct {
		Type   string `json:"Type"`
		Source string `json:"Source"`
	}
	if json.Unmarshal(out, &mounts) != nil {
		return ""
	}
	for _, m := range mounts {
		if m.Type == "bind" {
			return m.Source
		}
	}
	return ""
}

func streamZip(w http.ResponseWriter, root, zipName string) {
	w.Header().Set("Content-Type", "application/zip")
	w.Header().Set("Content-Disposition", `attachment; filename="`+zipName+`"`)
	zw := zip.NewWriter(w)
	defer zw.Close()

	_ = filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		rel, rerr := filepath.Rel(root, path)
		if rerr != nil || rel == "." {
			return nil
		}
		if isExcluded(rel) {
			if d.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}
		if d.IsDir() {
			return nil
		}
		f, oerr := os.Open(path)
		if oerr != nil {
			return nil
		}
		defer f.Close()
		hw, herr := zw.Create(filepath.ToSlash(rel))
		if herr != nil {
			return nil
		}
		_, _ = io.Copy(hw, f)
		return nil
	})
}

func (e Export) PM2(w http.ResponseWriter, r *http.Request) {
	ctx, cancel := context.WithTimeout(r.Context(), 15*time.Second)
	defer cancel()
	name := chi.URLParam(r, "name")
	path := e.pm2Path(ctx, name)
	if _, err := os.Stat(path); path == "" || err != nil {
		httpx.JSON(w, http.StatusNotFound, map[string]interface{}{
			"success": false, "error": "Project path not found. Make sure the PM2 process is running."})
		return
	}
	streamZip(w, path, name+"-"+timestamp()+".zip")
}

func (e Export) Docker(w http.ResponseWriter, r *http.Request) {
	ctx, cancel := context.WithTimeout(r.Context(), 15*time.Second)
	defer cancel()
	id := chi.URLParam(r, "id")
	path := e.dockerMount(ctx, id)
	if path == "" {
		httpx.JSON(w, http.StatusNotFound, map[string]interface{}{
			"success": false, "error": "No bind mount found or path doesn't exist. Container might be using volumes instead of bind mounts."})
		return
	}
	short := id
	if len(short) > 12 {
		short = short[:12]
	}
	streamZip(w, path, "docker-"+short+"-"+timestamp()+".zip")
}

func timestamp() string { return time.Now().Format("20060102150405") }
